tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,388 @@
+}
+
+// Test return-by-reference char fields by comparing to analytic solution
+template <size_t Dim, typename Solution>
+void test_characteristic_fields_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound) noexcept {
+  // Set up grid
+  Mesh<Dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+                 Spectral::Quadrature::GaussLobatto};
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-dependent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<Dim>, ScalarWave::Psi>{});
+  // Get ingredients
+  const size_t n_pts = x.begin()->size();
+  const auto gamma_2 = make_with_value<Scalar<DataVector>>(x, 0.1);
+  const auto& psi = get<ScalarWave::Psi>(vars);
+  const auto& phi = get<ScalarWave::Phi<Dim>>(vars);
+  const auto& pi = get<ScalarWave::Pi>(vars);
+  // Outward 3-normal to the surface on which characteristic fields are needed
+  auto unit_normal_one_form =
+      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(x, 1.);
+  const auto norm_of_one_form = magnitude(unit_normal_one_form);
+  for (size_t i = 0; i < Dim; ++i) {
+    unit_normal_one_form.get(i) /= get(norm_of_one_form);
+  }
+
+  // Compute characteristic fields locally
+  Scalar<DataVector> phi_dot_normal{DataVector(n_pts, 0.)};
+  for (size_t i = 0; i < Dim; ++i) {
+    get(phi_dot_normal) += unit_normal_one_form.get(i) * phi.get(i);
+  }
+
+  tnsr::i<DataVector, Dim, Frame::Inertial> phi_dot_projection_tensor{
+      DataVector(n_pts, 0.)};
+  for (size_t i = 0; i < Dim; ++i) {
+    phi_dot_projection_tensor.get(i) =
+        phi.get(i) - unit_normal_one_form.get(i) * get(phi_dot_normal);
+  }
+
+  const auto& upsi = psi;
+  const auto& uzero = phi_dot_projection_tensor;
+  const Scalar<DataVector> uplus{get(pi) + get(phi_dot_normal) -
+                                 get(gamma_2) * get(psi)};
+  const Scalar<DataVector> uminus{get(pi) - get(phi_dot_normal) -
+                                  get(gamma_2) * get(psi)};
+
+  // Check that locally computed fields match returned ones
+  const auto uvars = ScalarWave::CharacteristicFieldsCompute<Dim>::function(
+      gamma_2, psi, pi, phi, unit_normal_one_form);
+
+  const auto& upsi_from_func = get<ScalarWave::Tags::UPsi>(uvars);
```

Comment:
I suggest removing `_from_func` suffix after renaming the expected results

Reply:
changed everywhere